src/eth1/impl/ethers.ts

```diff
@@ -58,55 +68,85 @@ export class EthersEth1Notifier extends EventEmitter implements Eth1Notifier {
+    );
     if (index !== this.depositCount) {
       logger.warn(`Validator deposit with index=${index} received out of order.`);
       // deposit processed out of order
       return;
     }
     this.depositCount++;
-
     const data: DepositData = deserialize(dataBuf, DepositData);
-
-    // TODO: Add deposit to merkle trie/db
-    this.emit('deposit', data, index);
+    const deposit: Deposit = {
+      index: index,
+      //TODO: calculate proof
+      proof: [],
+      data
+    };
+    //after genesis stop storing in genesisDeposit bucket
+    if(!this.genesisBlockHash) {
+      await this.db.setGenesisDeposit(deposit);
+    }
+    this.emit('deposit', deposit);
   }
 
-  public async processEth2GenesisLog(depositRootHex: string, depositCountHex: string, timeHex: string, event: ethers.Event): Promise<void> {
+  public async processEth2GenesisLog(
+    depositRootHex: string,
+    depositCountHex: string,
+    timeHex: string,
+    event: ethers.Event
+  ): Promise<void> {
     const depositRoot = Buffer.from(depositRootHex.substr(2), 'hex');
     const depositCount = Buffer.from(depositCountHex.substr(2), 'hex').readUIntLE(0, 6);
     const time = new BN(Buffer.from(timeHex.substr(2), 'hex').readUIntLE(0, 6));
     const blockHash = Buffer.from(event.blockHash.substr(2), 'hex');
+    this.genesisBlockHash = event.blockNumber;
     logger.info(`Received Eth2Genesis event. blockNumber=${event.blockNumber}, time=${time}`);
-    // TODO: Ensure the deposit root is the same that we've stored
 
     const genesisEth1Data: Eth1Data = {
       depositRoot,
       blockHash,
       depositCount,
     };
-
-    this.chainStarted = true;
-    this.emit('eth2genesis', time, await this.genesisDeposits(), genesisEth1Data);
+    const genesisDeposits = await this.genesisDeposits();
+    this.emit('eth2genesis', time, genesisDeposits, genesisEth1Data);
+    //from now on it will be kept in BeaconBlock
+    await this.db.deleteGenesisDeposits(genesisDeposits);
   }
 
   public async genesisDeposits(): Promise<Deposit[]> {
-    return [];
+    return this.db.getGenesisDeposits();
   }
 
-  public async latestBlockHash(): Promise<bytes32> {
+  public latestBlockHash(): bytes32 {
     return this._latestBlockHash;
   }
 
   public async depositRoot(): Promise<bytes32> {
-    return Buffer.alloc(32);
+    const depositRootHex = await this.contract.get_deposit_root();
+    return Buffer.from(depositRootHex.substr(2), 'hex');
+  }
+
+  private async initContract(): Promise<void> {
+    const address = this.opts.depositContract.address;
+    const abi = this.opts.depositContract.abi;
+    if (!(await this.contractExists(address))) {
+      throw new Error(`There is no deposit contract at given address: ${address}`);
+    }
+    try {
+      this.contract = new ethers.Contract(address, abi, this.provider);
+    } catch (e) {
+      throw new Error('Eth1 deposit contract not found! Probably wrong eth1 rpc url');
```

Comment:
Will this error be thrown?
I haven't been able to get `new ethers.Contract` to throw an error using the default provider. It looks there's a `contract.addressPromise` that can be rejected though.

Reply:
It won't since there is a bug in ethers.js where sending "TbD" as error ends up in promise rejection, I'm hopping they will fix it. It's an old code here, I just moved it into separated directory to be more clear what is what.